### ISSUE TITLE
Application freezes when attempting to display GIF, video or image

### ISSUE THREAD

**Issue by stonest** [eneshecan/whatsapp-for-linux#181]
**Description**
When opening a chat containing GIFs in the chat history, the application freezes and I am unable to do anything in it until I close and reopen the application. The GIFs show the spinning circle indicating that they are attempting to load, however, I cannot click anything in the application (other chats, try to cancel the GIF from loading). I can only restart the application.

I'm able to replicate this problem on two different machines. 
Running whatsapp via the console doesn't show any errors when I click to a chat with a GIF

**To Reproduce**
1. Open whatsapp
2. Click on a chat containing a GIF
3. See the application become unresponsive

**Expected behavior**
GIFs should be loaded without issue and without causing the application to become unresponsive

**Environment**
 - Distro - Arch Linux
 - Package yay
 - Desktop Environment or Window Manager KDE Plasma

**Comment by eneshecan** [comment]
Hi @stonest. Thank you for reporting. Yes, I also recently experienced this with videos and gifs. I saw some pipewire logs while this was happening. Needs to be investigated.

**Comment by eneshecan** [comment]
Somehow running with `WEBKIT_DISABLE_COMPOSITING_MODE=1` fixes but I am not sure how... Could you please check? @stonest

**Comment by stonest** [comment]
@eneshecan this appears to have worked.
This may help, I get these messages in the console now when I open a chat with GIFs:
```<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.615: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4677:gst_pad_push_data:<qtdemux0:video_0> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.615: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4416:gst_pad_chain_data_unchecked:<multiqueue0:sink_0> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.615: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4677:gst_pad_push_data:<multiqueue0:src_0> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.615: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4416:gst_pad_chain_data_unchecked:<h264parse0:sink> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.615: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4677:gst_pad_push_data:<h264parse0:src> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.615: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4416:gst_pad_chain_data_unchecked:<capsfilter0:sink> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.615: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4677:gst_pad_push_data:<capsfilter0:src> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.615: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4416:gst_pad_chain_data_unchecked:<avdec_h264-0:sink> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.677: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4677:gst_pad_push_data:<qtdemux1:video_0> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.677: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4416:gst_pad_chain_data_unchecked:<multiqueue1:sink_0> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.677: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4677:gst_pad_push_data:<multiqueue1:src_0> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.677: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4416:gst_pad_chain_data_unchecked:<h264parse1:sink> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.677: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4677:gst_pad_push_data:<h264parse1:src> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.677: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4416:gst_pad_chain_data_unchecked:<capsfilter1:sink> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.677: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4677:gst_pad_push_data:<capsfilter1:src> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.677: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4416:gst_pad_chain_data_unchecked:<avdec_h264-1:sink> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.679: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4677:gst_pad_push_data:<h264parse1:src> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.679: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4416:gst_pad_chain_data_unchecked:<capsfilter1:sink> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.679: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4677:gst_pad_push_data:<capsfilter1:src> Got data flow before segment event
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: 
<13>Mar 30 14:00:39 whatsapp-for-linux[45781]: (WebKitWebProcess:45792): GStreamer-WARNING **: 14:01:02.679: ../gstreamer/subprojects/gstreamer/gst/gstpad.c:4416:gst_pad_chain_data_unchecked:<avdec_h264-1:sink> Got data flow before segment event```

**Comment by eneshecan** [comment]
Might be a bug in `GStreamer` related to hardware acceleration. Disabling hardware acceleration in WebKit also fixes.

**Comment by eneshecan** [comment]
I just opened up the application today and saw that issue is gone without any workarounds. Can someone else confirm this?

**Comment by stonest** [comment]
Unfortunately it hasn't worked for me, it is still freezing if i try to open an imag, using `WEBKIT_DISABLE_COMPOSITING_MODE=1` still works

**Comment by ghostface** [comment]
Same problem here with the same workaround (WEBKIT_DISABLE_COMPOSITING_MODE=1) working.

I am using sway (wayland) btw.

**Comment by mitchellkrogza** [comment]
> Same problem here with the same workaround (WEBKIT_DISABLE_COMPOSITING_MODE=1) working.
> 
> I am using sway (wayland) btw.



> Somehow running with `WEBKIT_DISABLE_COMPOSITING_MODE=1` fixes but I am not sure how... Could you please check? @stonest

Thanks this fix works for me

**Comment by mitchellkrogza** [comment]
This fix works for stopping it freezing but if you hold your right arrow down while scrolling forward through something you have typed before sending, after 2 seconds of holding down the right arrow the entire program crashes and renders a grey screen. Its a pity but this app is becoming quite unusable.

**Comment by GlowingScrewdriver** [comment]
> This fix works for stopping it freezing but if you hold your right arrow down while scrolling forward through something you have typed before sending, after 2 seconds of holding down the right arrow the entire program crashes and renders a grey screen. Its a pity but this app is becoming quite unusable. Also freezes trying to send a GIF

I have see the same issue. On a bit of investigation using the developer console, I find the following error:

```
[Error] SecurityError: Attempt to use history.pushState() more than 100 times per 30 seconds — bootstrap_main.c7fae15c2be134004f59.js:13:1741782
	ss (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:25:82574)
	(anonymous function) (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:25:84827)
	co (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:25:48036)
	Va (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:25:70670)
	Tu (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:25:98765)
	ku (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:25:98693)
	Su (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:25:98556)
	mu (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:25:95543)
	mu
	(anonymous function) (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:25:45214)
	(anonymous function) (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:49:4012)
	Wi (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:25:45160)
	$i (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:25:45095)
	De (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:25:116079)
	Zt (vendor1~bootstrap_qr.010e0b1e6789f30ac249.js:25:23756)
	Zt
```
Repeated the process three times, and this same error message gets displayed in the webview console every time, at the same instant that the screen goes blank.

**Comment by eneshecan** [comment]
> > So, possible workarounds I have found:
> > 
> > * Running the application with `WEBKIT_DISABLE_COMPOSITING_MODE=1`
> 
> Why not implement this in the desktop launcher?

Well, we can do that but I didn't want to touch the default hardware acceleration policy set by WebKit which might be helpful for using the graphics card for those not having this issue.

**Comment by dlight** [comment]
>     * Running the application with `WEBKIT_DISABLE_COMPOSITING_MODE=1`

This worked for me on amdgpu on Linux, thanks. However, I never got the " (WebKitWebProcess:45792): GStreamer-WARNING" message on terminal, the application just hang whenever I opened a conversation with a GIF (even if the GIF was scrolled up, outside the viewport)

(if relevant I'm on Arch Linux, GPU Radeon RX Vega 7 Renoir, driver amdgpu, running KDE 5.25 on X11, kernel 5.18.5)

**Comment by dlight** [comment]
Maybe it's possible to detect this hang in code (maybe by trying to interact with UI elements programmatically and failing to obtain a response), and write to some file on `~/.cache/whatsapp-for-linux` something that makes the next start disable compositing.

Sounds hacky but.

**Comment by eneshecan** [comment]
Yes, it is possible to detect this.

One thing we can do is ask the user if they want to kill and reload the process if the page is not responding. I already did this locally and works fine. However, I don't want to go for any workarounds since this is most probably an upstream issue as I mentioned before. What do you guys say?

**Comment by GlowingScrewdriver** [comment]
Well... How about also having a settings menu option? A user might  experience a crash once or twice, and then decide they just want to launch the app without hardware acceleration.

**Comment by eneshecan** [comment]
Okay. I will do that as an option then. I will also create a `Preferences` window to move related settings into there instead of showing them in the header menu. See #205

**Comment by eneshecan** [comment]
Alright. I pushed 0cf9e56 to recover from this situation.

**Comment by eneshecan** [comment]
Here I implemented this setting as well: f7eb282
